### PR TITLE
Update course awards to have validation and use demo service

### PR DESCRIPTION
**Pull request by quinnbudan**
Also updated the validation service to test for non-negative real numbers, and updated formatting to match the standard we've set.

### CODE REVIEW

File: src/components/add-awards-dialog.js [hunk 1]
```diff
@@ -7,11 +7,13 @@ import { BaseMixin } from '../mixins/base-mixin';
 export const awardsDialogId = 'add-award-dialog';
 const DONE_ACTION = 'done';
 const CANCEL_ACTION = 'cancel';
+const ORG_UNIT_ID = 'tmp_org_unit_id';
```
Comment (BraydonKains) [inline src/components/add-awards-dialog.js:10]: nit: With real data this will be a number, so might as well make this a number too.

File: src/components/add-awards-dialog.js [hunk 2]
```diff
@@ -7,11 +7,13 @@ import { BaseMixin } from '../mixins/base-mixin';
...
 
 class AddAwardsDialog extends BaseMixin(LitElement) {
 	static get properties() {
 		return {
-			availableAwards: { type: Object }
+			opened: { type: Boolean },
```
Comment (BraydonKains) [inline src/components/add-awards-dialog.js:15]: nit: We should pick a standard format for defining properties. Our choices are one line if possible or to always split it into more lines. I lean to the latter, but we can discuss offline.

File: src/components/add-awards-dialog.js [hunk 3]
```diff
@@ -28,35 +30,52 @@ class AddAwardsDialog extends BaseMixin(LitElement) {
...
-				id: '789',
-				name: 'Available Award',
-				imgPath: '../../images/example_award.png',
-				type: 'Badge',
-				credits: 5,
-				hiddenUntilEarned: false,
-			}
-		];
+		this.opened = false;
+		this.availableAwards = [];
```
Comment (BraydonKains) [inline src/components/add-awards-dialog.js:34]: Q: Was lit-analyzer okay with you not initializing this with `Array()`? Louis had to do that.

File: src/components/add-awards-dialog.js [hunk 4]
```diff
@@ -28,35 +30,52 @@ class AddAwardsDialog extends BaseMixin(LitElement) {
...
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+		this._fetchAvailableAwards();
+	}
+
+	async _fetchAvailableAwards() {
+		const { awards } = await window.AwardService.getAvailableAwards();
```
Comment (BraydonKains) [inline src/components/add-awards-dialog.js:43]: we stan es6 destructuring

File: src/components/add-awards-dialog.js [hunk 5]
```diff
@@ -28,35 +30,52 @@ class AddAwardsDialog extends BaseMixin(LitElement) {
...
+	}
+
+	async _handleDialogClosed(event) {
+		const { detail: { action } } = event;
+
+		// Get the selection info and then clear the selected items
+		const d2lList = this.shadowRoot.querySelector('d2l-dialog d2l-list');
+		const { keys } = d2lList.getSelectionInfo();
+		const items = this.shadowRoot.querySelectorAll('d2l-dialog d2l-list d2l-list-item');
+		items.forEach(item => item.setSelected(false, true)); // selected, suppressEvent
```
Comment (BraydonKains) [inline src/components/add-awards-dialog.js:63]: Q: Does `setSelected` not support named args?

File: src/components/add-awards-dialog.js [hunk 6]
```diff
@@ -28,35 +30,52 @@ class AddAwardsDialog extends BaseMixin(LitElement) {
...
+		this.dispatchEvent(doneClickedEvent);
+	}
+
+	async _handleDialogClosed(event) {
+		const { detail: { action } } = event;
+
+		// Get the selection info and then clear the selected items
+		const d2lList = this.shadowRoot.querySelector('d2l-dialog d2l-list');
+		const { keys } = d2lList.getSelectionInfo();
+		const items = this.shadowRoot.querySelectorAll('d2l-dialog d2l-list d2l-list-item');
```
Comment (BraydonKains) [inline src/components/add-awards-dialog.js:62]: Q: Why didn't just selecting `d2l-list` and `d2l-list-items` work?

File: src/components/add-awards-dialog.js [hunk 7]
```diff
@@ -28,35 +30,52 @@ class AddAwardsDialog extends BaseMixin(LitElement) {
...
-			const d2lList = d2lDialog.querySelector('d2l-list');
-			const { keys } = d2lList.getSelectionInfo();
 			const awards = this.availableAwards.filter(award => keys.includes(award.id));
-
-			return awards;
+			if (awards.length !== 0) {
+				await window.AwardService.addAwardsToOrgUnit({ awards, orgUnitId: ORG_UNIT_ID });
+				this._fireCustomEvent();
+			} else {
+				console.log('Not calling API because no awards were selected');
```
Comment (BraydonKains) [inline src/components/add-awards-dialog.js:71]: This probably should be deleted

File: src/services/validation-service.js [hunk 8]
```diff
@@ -1,4 +1,10 @@
 export class ValidationService {
+	static isNonNegativeNumber(str) {
+		if (!str) return false;
+		const nonNegativeNumberPattern = /^(\d+)*(\.\d+)*$/;
```
Comment (BraydonKains) [inline src/services/validation-service.js:4]: Still amazed there isn't a better way to do this. *shakes fist* javascriiipt

File: src/components/add-awards-dialog.js [hunk 9]
```diff
@@ -28,35 +30,52 @@ class AddAwardsDialog extends BaseMixin(LitElement) {
...
 	}
 
-	async openAndGetItems() {
-		const d2lDialog = this.shadowRoot.querySelector('d2l-dialog');
-		const action = await d2lDialog.open();
+	_fireCustomEvent() {
+		const doneClickedEvent = new CustomEvent('d2l-add-awards-dialog-done-clicked', {
+			detail: {},
+			bubbles: true,
+			composed: true
```
Comment (louiscoste) [inline src/components/add-awards-dialog.js:51]: What are these parameters for? Why are they not in detail?

File: src/components/add-awards-dialog.js (hunk 5 above)
Comment (quinnbudan) [reply src/components/add-awards-dialog.js:63]: If you're thinking about named args the way I've done them in the demo service (using an object), then no. There might be another way that I don't know of. In other places it's used hardcoded booleans are passed in with no names as well

File: src/services/validation-service.js (hunk 8 above)
Comment (quinnbudan) [reply src/services/validation-service.js:4]: Yeah you can combine functions like `isNan()` and `Number()` and `parseFloat()` but this felt more reliable and honestly simpler...

File: src/components/add-awards-dialog.js (hunk 2 above)
Comment (quinnbudan) [reply src/components/add-awards-dialog.js:15]: Since Louis also always split into more lines, I'll update to split these into more lines (already making updates)

File: src/components/add-awards-dialog.js (hunk 3 above)
Comment (quinnbudan) [reply src/components/add-awards-dialog.js:34]: It actually does pass lit-analyzer, however I had other lint errors

File: src/components/add-awards-dialog.js (hunk 1 above)
Comment (quinnbudan) [reply src/components/add-awards-dialog.js:10]: updating

File: src/services/award-service.js [hunk 10]
```diff
@@ -1,9 +1,25 @@
 export class AwardService {
+	static async addAwardsToOrgUnit() {
+		return {};
+	}
+
+	static async deleteAward() {
+		return {};
+	}
+
+	static async getAvailableAwards() {
```
Comment (louiscoste) [inline src/services/award-service.js:10]: I thought this would end up being done by calling getAwards and passing in some parameters. Unless there are some strong opinions about doing it that way

File: src/components/add-awards-dialog.js (hunk 6 above)
Comment (quinnbudan) [reply src/components/add-awards-dialog.js:62]: It does work! I was just being overly specific, updating to use just those

File: src/components/add-awards-dialog.js (hunk 9 above)
Comment (quinnbudan) [reply src/components/add-awards-dialog.js:51]: These are custom event parameters defined [here](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent), essentially it specifies that this event can bubble up so that elements higher in the DOM (and outside the shadow DOM) can listen for it. Not entirely necessary to set these to true but it doesn't hurt anything as far as I know. However, I think I'll set them to false anyway.

File: src/services/award-service.js (hunk 10 above)
Comment (quinnbudan) [reply src/services/award-service.js:10]: Maybe, but it's likely an API call to the awards API rather than the associations API. Even if it is a call to the associations API, it still may be nice to expose it via a clearer name in the service. I wouldn't say my opinion is strong though, I'm happy to remove this and use `getAwards()` only

Comment (quinnbudan) [reply src/services/award-service.js:10]: Thinking: maybe we should rename `getAwards()` to `getAssociations()`?

File: src/components/add-awards-dialog.js (hunk 7 above)
Comment (quinnbudan) [reply src/components/add-awards-dialog.js]: deleted

File: src/components/add-awards-dialog.js (hunk 5 above)
Comment (BraydonKains) [reply src/components/add-awards-dialog.js:66]: It's just generally if a comment like this is necessary to describe what's happening it's a smell that something should be done to make the code more readable. Not worth holding the PR back.

File: src/services/award-service.js (hunk 10 above)
Comment (BraydonKains) [reply src/services/award-service.js:10]: Awards and Associations are two different things, so if `getAwards` is currently reflecting associations then could use a rename. We'll come back around to this when the real APIs are hooked up.
